cf_remote/commands.py

```diff
@@ -475,18 +475,56 @@ def uninstall(hosts):
         errors += uninstall_host(host)
     return errors
 
-def deploy(hubs, directory):
-    assert(directory.endswith("/masterfiles"))
-    assert(os.path.isfile(directory + "/autogen.sh"))
+def deploy_tarball(hubs, tarball):
+    assert os.path.isfile(tarball)
+
+    if not tarball.endswith((".tgz", ".tar.gz")):
+        log.error("The masterfiles directory must be in a gzipped tarball (.tgz or .tar.gz)")
+        return 1
+
+    errors = 0
+    for hub in hubs:
+        errors += deploy_masterfiles(hub, tarball)
+    return errors
+
+def deploy(hubs, masterfiles):
+    if masterfiles.startswith(("http://", "https://")):
+        urls = [masterfiles]
+        paths = _download_urls(urls)
+        assert len(paths) == 1
+        masterfiles = paths[0]
+        log.debug(f"Deploying downloaded: {masterfiles}")
+    else:
+        masterfiles = os.path.abspath(os.path.expanduser(masterfiles))
+        log.debug(f"Deploy path expanded to: {masterfiles}")
+
+    if masterfiles.endswith("/"):
+        masterfiles = masterfiles[0:-1]
+
+    if os.path.isfile(masterfiles):
+        return deploy_tarball(hubs, masterfiles)
+
+    if not os.path.isdir(masterfiles):
+        log.error(f"'{masterfiles}' must be a directory")
+        return 1
+
+    directory = masterfiles
+
+    if not directory.endswith("/masterfiles"):
+        log.error("The masterfiles directory to deploy must be called 'masterfiles'")
+        return 1
+    if not os.path.isfile(directory + "/autogen.sh"):
```

Comment:
Here, I'd like two options: autogen.sh -> promises.cf, or if promises.cf is already present (such as in an "installed" policy set) don't run autogen.sh at all and allow it to proceed.

Reply:
But I definitely do want it to run autogen again and again, because I will be editing `promises.cf.in`, and want my `promises.cf` to be updated.

Reply:
That's fine actually. If `autogen.sh` is there, run it. If not, let it pass and deploy if `promises.cf` is present since that would be my case of "installed policy".

Reply:
Cool, I have added that in the latest commit.